### PR TITLE
Add warning if there are several instances of styled-components on the page

### LINKED ISSUE #1011: Add warning on StyleSheet rehydration when styled-components is duplicated
When styled-components is duplicated in the client runtime, the rehydration will either throw a rather cryptic error on rehydration, or stop injecting new, dynamic styles.

Instead of failing this catastrophically we should add a warning that notifies the user of this mistake.

cc @geelen @mxstbr 

Reproduction: https://glitch.com/edit/#!/sc-duplicate-module-reproduction

This issue replaces #992 and #987

### PR DESCRIPTION
**Pull request by gribnoysup**
Closes #1011 and also a follow-up to https://github.com/styled-components/styled-components/issues/1395#issuecomment-355268773

I think that even if the https://github.com/styled-components/styled-components/issues/1102 will allow users to solve major issues with having several instances of `s-c` on the page, there is still a factor of having several modules loaded to the client for no good reason as almost all cases described in several tickets related to this issue (https://github.com/styled-components/styled-components/issues/992, https://github.com/styled-components/styled-components/issues/987) could be solved with just one instance of `s-c` on the page. 

If you feel that this warning is too demanding of the end users of the library, feel free to close the PR (and the issue then should be also closed, probably)

### CODE REVIEW

**Comment by gribnoysup** [comment]
@mxstbr yep, this is a good idea! It seems that there are several scenarios when people encounter this problem: 

1) Using a monorepo-like setup for the project (https://github.com/styled-components/styled-components/issues/1395#issuecomment-355083444)
2) Dynamic imports bundles several styled-components in separate child chunks (https://github.com/styled-components/styled-components/issues/992#issuecomment-315167419)
3) Duplication of styled-components in the bundle because of the `node_modules` dependencies (https://github.com/styled-components/styled-components/issues/987#issuecomment-316318051)

Regarding the third one, am I assuming correctly that for the library authors we should recommend marking `styled-components` as a peer dependency and not putting it into the bundle?

**Comment by mxstbr** [comment]
Yep, that sounds reasonable. Wanna PR over at https://github.com/styled-components/styled-components-website?

**Comment by gribnoysup** [comment]
Sure! I'll be glad to help with this

**Comment by RahulDole** [comment]
Eagerly waiting for this update. Any estimate on when will this be done?

**Comment by gribnoysup** [comment]
@RahulDole need to finish with the docs first, working on this right now 😉 This PR doesn't solve the problem though, this one just adds a warning to the console. Do you mind me asking why are you waiting for this so much?

**Comment by RahulDole** [comment]
I am waiting because I have the same problem of duplicate instances. I have this in my `webpack.config`, `entry: {
    'app': './src/index.js',
    'try': './src/index.try.js'
  }`, so both these files get their own version of styled component. Hence, the `try.js` doesn't render styles on its styled components. For now I'm making do with manual inline styling as its just one file where I have to make this change.

**Comment by philpl** [comment]
@RahulDole you should set up a commons chunk in webpack ;) there’s a plugin for it and you can either make it extract modules automatically or prespecify a vendor entry point.

Btw please ask for help on the original issue and not the PR 😊

**Comment by RahulDole** [comment]
@philpl Thanks for the help. Sorry I got lost in all the different but related threads on this issue. Then I bumped over this PR and it looked like this is the final PR to solve the issue or throw a warning which has guidance on an alternate solution.
Is [this](https://github.com/styled-components/styled-components/pull/1102) the issue link you want me to write in?

**Comment by gribnoysup** [comment]
Docs draft is ready for review! Also merged master and updated the `CHANGELOG.md`

**Comment by mxstbr** [comment]
@gribnoysup here is a short link! `https://s-c.sh/2BAXzed`

**Comment by gribnoysup** [comment]
@mxstbr thank you, changed the warning message! Should we cover warning messages with tests? 🤔

**Comment by gribnoysup** [comment]
As the `StyleSheet.target` support (https://github.com/styled-components/styled-components/pull/1491) is finally in the `master` 🎉  I wanted to discuss the fate of this PR. 

The issue of having several instances of `styled-components` on one page is solvable right now, although as I understand the main purpose of having `target` is not exactly to bundle several `styled-components` modules in your app 😉 and detecting the problem would still be not that easy in some cases.

So what should we do with this? Discard this change? Update detection logic to focus more on using the same `target` on one page, and not just several instances? Leave it as is? Any other ideas?

Just want to know what do you think about this! 

/cc @mxstbr @kitten (nice new username, btw 😄) @probablyup

**Comment by kitten** [comment]
@gribnoysup As `target` is the "output" of our style tags, the "input" i.e. SSR rehydration is still problematic when two styled-components instances are present 😢 So this is still relevant

cc @mxstbr ready to go I'd say

**Comment by mxstbr** [comment]
Thanks for pushing this through @gribnoysup! :tada:

File: src/index.js [hunk 1]
```diff
@@ -30,13 +30,32 @@ if (
...
+  // eslint-disable-next-line no-console
   console.warn(
     "It looks like you've imported 'styled-components' on React Native.\n" +
       "Perhaps you're looking to import 'styled-components/native'?\n" +
       'Read more about this at https://www.styled-components.com/docs/basics#react-native'
   )
 }
 
+/* Warning if there are several instances of styled-components */
+if (process.env.NODE_ENV !== 'production' && typeof window !== 'undefined') {
```
Comment (oliviertassinari) [inline src/index.js]: Any specific reason for ignoring the server side case (`typeof window !== 'undefined'`)? Was it linked to HMR issues or the lack of a good global ponyfill?
```js
// https://github.com/zloirock/core-js/issues/86#issuecomment-115759028
export default (typeof window != 'undefined' && window.Math == Math
  ? window
  : typeof self != 'undefined' && self.Math == Math
  ? self
  : Function('return this')());
```

**Pull request merged**